stdlib/json.gr

```diff
@@ -0,0 +1,2261 @@
+  from WasmI32 use {
+    and as (&),
+    eq as (==),
+    ne as (!=),
+    shl as (<<),
+    shrS as (>>),
+  }
+
+  let ptr = WasmI32.fromGrain(value)
+  if ((ptr & 1n) != 0n) {
+    // FIXME expose runtime/numbers.untagSimple?
+    let asWasmI32 = ptr >> 1n
+    printNumberWasmI32(asWasmI32, buffer)
+    None
+  } else if ((ptr & 7n) == Tags._GRAIN_GENERIC_HEAP_TAG_TYPE) {
+    let tag = WasmI32.load(ptr, 0n)
+    match (tag) {
+      t when t == Tags._GRAIN_BOXED_NUM_HEAP_TAG => {
+        let numberTag = WasmI32.load(ptr, 4n)
+        match (numberTag) {
+          t when t == Tags._GRAIN_INT64_BOXED_NUM_TAG => {
+            let asWasmI64 = WasmI64.load(ptr, 8n)
+            // FIXME magic numbers
+            printNumberWasmI64(asWasmI64, buffer)
+            None
+          },
+          t when t == Tags._GRAIN_BIGINT_BOXED_NUM_TAG => {
+            Buffer.addString(BI.bigIntToString10(ptr), buffer)
+            None
+          },
+          t when t == Tags._GRAIN_RATIONAL_BOXED_NUM_TAG => {
+            from WasmF64 use { div as (/) }
+            // As a compromnise, convert the rational into a float
+            // and then print that. This means there can be
+            // precision loss even for values that could be printed
+            // exactly in decimal notation, but doing otherwise
+            // could both compromise perfomance and confuse
+            // developers as most JSON libraries in other languages
+            // likely only deal with integers and floats.
+            // Note that the ECMA-404 standard only specifies how
+            // numbers should be formatted as text (in decimal
+            // notation), so we could go the extra mile and at
+            // least for some numbers try printing the exact
+            // number. For example those that have a power of 10
+            // in the denominator.
+            let asWasmF64 = WasmF64.convertI32S(
+              Numbers.boxedRationalNumerator(ptr)
+            ) /
+              WasmF64.convertI32S(Numbers.boxedRationalDenominator(ptr))
+
+            printNumberWasmF64(asWasmF64, buffer)
+          },
+          t when t == Tags._GRAIN_FLOAT64_BOXED_NUM_TAG => {
+            let asWasmF64 = WasmF64.load(ptr, 8n)
+            // FIXME magic numbers
+            printNumberWasmF64(asWasmF64, buffer)
+          },
+          _ => {
+            throw UnknownNumberTag
+          },
+        }
+      },
+      _ => {
+        throw UnknownNumberTag
+      },
+    }
+  } else {
+    throw UnknownNumberTag
+  }
+}
+
+// Note that this function is not allocated with the JSONWriter because currently
+// Grain leaks memory each time a recursive closure is allocated. Not reallocating
+// it may also be a better thing to do.
+let rec printElementCompact =
+  (
+    json: JSON,
+    implHelper: JSONWriterCompactImplHelper,
+  ) => {
+  let buffer = implHelper.buffer
+  match (json) {
+    JSONNull => {
+      printNull(buffer)
+      None
```

Comment:
Since for stringifying there are a fixed number of possible things that could go wrong, perhaps instead of using `None` and `Some` for determining if there was an error a `WasmI32` could be used instead to reduce allocations as a performance optimization? Maybe `-1n` could mean no error, `0n` for `NaN not allowed ...`, `1n` for `-Infinity not allowed ...`, etc? Then the error message could be constructed at the toplevel `toString`?

Reply:
I tried to strike a balance between performant and functional/idiomatic code. For OK case, passing around the reference to `None` doesn't actually allocate anything and I didn't consider important optimizing the error code path. There's only the cost of reference counting compared to using a `WasmI32`, which would mostly go away with a tracing GC. I'm assuming all native Wasm GC implementations will use be tracing. Also back then we didn't have `@unsafe` and mixing `WasmI32` inside an otherwise safe code was not really an option. A function had to either be completely low level non GC code or fully managed with occasional calls to non GC functions.

Reply:
This was discussed a little bit more on the discord im not a big fann of usig the `WasmI32` type even now but we can use early `return` which should make thigs even more readable.